src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -1371,21 +1378,29 @@ private bool ShouldAsyncBeTreatedAsModifier(bool parsingStatementNotDeclaration)
 
         private static bool IsNonContextualModifier(SyntaxToken nextToken)
         {
-            return GetModifier(nextToken) != DeclarationModifiers.None && !SyntaxFacts.IsContextualKeyword(nextToken.ContextualKind);
+            return GetModifier(nextToken.Kind, contextualKind: SyntaxKind.None) != DeclarationModifiers.None;
         }
 
-        private bool IsPartialType()
+        private static bool CanFollowPartialKeywordOfPartialType(SyntaxToken token, out bool isPartialFollowedByType)
         {
-            Debug.Assert(this.CurrentToken.ContextualKind == SyntaxKind.PartialKeyword);
-            switch (this.PeekToken(1).Kind)
+            switch (token.Kind)
             {
                 case SyntaxKind.StructKeyword:
                 case SyntaxKind.ClassKeyword:
                 case SyntaxKind.InterfaceKeyword:
+                    isPartialFollowedByType = true;
                     return true;
+
+                default:
+                    isPartialFollowedByType = false;
+                    return IsNonContextualModifier(token);
```

Comment:
>return IsNonContextualModifier(token); [](start = 20, length = 38)

It is not clear to me why do we restrict next modifier to non-contextual one. If we only want to allow ```ref``` then we should check for ref. Otherwise, we probably should allow contextual modifiers as well. Regarding non-contextual modifiers, should we return true for ```partial public struct```?

Reply:
That's because in `partial partial;` the second token should not be parsed as a modifier.

To be clear, `IsNonContextualModifier` check if the token is any modifier excluding `async` and `partial`, so `partial public struct` is fine (we do have a test for `partial readonly public struct`).

Reply:
> That's because in partial partial; the second token should not be parsed as a modifier.

What is wrong with that? Why not parse it as a modifier, but report an error?

Reply:
Because it can be a valid code,
```cs
class partial {
  async async;
}
class async {
  partial partial;
}
```
If we parse it as a modifier, the resulting syntax node would be malformed and we get a false error.